lib/ObjWriter/typeBuilder.cpp

```diff
@@ -130,10 +132,91 @@ unsigned UserDefinedTypesBuilder::GetCompleteClassTypeIndex(
                  TypeIndex(), TypeIndex(), ClassFieldsDescriptor.Size,
                  ClassDescriptor.Name, ClassDescriptor.UniqueName);
   TypeIndex ClassIndex = TypeTable.writeKnownType(CR);
+
   if (ClassDescriptor.IsStruct == false) {
-    PointerRecord PointerToClass(ClassIndex, 0);
-    TypeIndex PointerIndex = TypeTable.writeKnownType(PointerToClass);
-    return PointerIndex.getIndex();
+    return GetPointerType(ClassIndex);
   }
   return ClassIndex.getIndex();
 }
+
+unsigned UserDefinedTypesBuilder::GetArrayTypeIndex(
+    ClassTypeDescriptor ClassDescriptor, ArrayTypeDescriptor ArrayDescriptor) {
+  TypeIndex ElementTypeIndex = TypeIndex(ArrayDescriptor.ElementType);
+  TypeIndex IndexType = TypeIndex(SimpleTypeKind::Int32);
+
+  FieldListRecordBuilder FLBR(TypeTable);
+  FLBR.begin();
+
+  unsigned Offset = 0;
+  unsigned FieldsCount = 0;
+
+  assert(ClassDescriptor.BaseClassId != 0);
+  AddBaseClass(FLBR, ClassDescriptor.BaseClassId);
+  FieldsCount++;
+  Offset += TargetPointerSize;
+
+  MemberAccess Access = MemberAccess::Public;
+
+  DataMemberRecord CountDMR(Access, IndexType, Offset, "count");
+  FLBR.writeMemberType(CountDMR);
+  FieldsCount++;
+  Offset += TargetPointerSize;
+
+  if (ArrayDescriptor.IsMultiDimensional == 1) {
+    for (unsigned i = 0; i < ArrayDescriptor.Rank; ++i) {
+      std::stringstream ss;
+      ss << "length" << i;
+      char *LengthName = new char[ss.gcount() + 1];
```

Comment:
This looks like a memory leak. You can probably just use ss.str().c_str() here to pass a char * through to the debug record serializer. Most string passing is done in LLVM with `llvm::StringRef` which might be more appropriate here, too.